ext/phar/dirstream.c

```diff
@@ -166,118 +150,97 @@ static int phar_compare_dir_name(Bucket *f, Bucket *s)  /* {{{ */
  * files in a phar and retrieving its relative path.  From this, construct
  * a list of files/directories that are "in" the directory represented by dir
  */
-static php_stream *phar_make_dirstream(char *dir, HashTable *manifest) /* {{{ */
+static php_stream *phar_make_dirstream(const char *dir, size_t dirlen, const HashTable *manifest) /* {{{ */
 {
 	HashTable *data;
-	size_t dirlen = strlen(dir);
-	char *entry, *found, *save;
-	zend_string *str_key;
-	size_t keylen;
-	zend_ulong unused;
+	char *entry;
 
 	ALLOC_HASHTABLE(data);
 	zend_hash_init(data, 64, NULL, NULL, 0);
 
 	if ((*dir == '/' && dirlen == 1 && (manifest->nNumOfElements == 0)) || (dirlen >= sizeof(".phar")-1 && !memcmp(dir, ".phar", sizeof(".phar")-1))) {
 		/* make empty root directory for empty phar */
 		/* make empty directory for .phar magic directory */
-		efree(dir);
 		return php_stream_alloc(&phar_dir_ops, data, NULL, "r");
 	}
 
-	zend_hash_internal_pointer_reset(manifest);
-
-	while (FAILURE != zend_hash_has_more_elements(manifest)) {
-		if (HASH_KEY_NON_EXISTENT == zend_hash_get_current_key(manifest, &str_key, &unused)) {
-			break;
-		}
-
-		keylen = ZSTR_LEN(str_key);
+	zend_string *str_key;
+	ZEND_HASH_MAP_FOREACH_STR_KEY(manifest, str_key) {
+		size_t keylen = ZSTR_LEN(str_key);
 		if (keylen <= dirlen) {
 			if (keylen == 0 || keylen < dirlen || !strncmp(ZSTR_VAL(str_key), dir, dirlen)) {
```

Comment:
I don't exactly understand what these conditions are doing, but it _feels_ like it can be refactored to use one of the zend_string APIs

Reply:
I would have to dig deeper to understand how this works, but let's cut the PR here.